### PR TITLE
v2.5.0: PRISM Dashboard TUI, CLI Snapshot, Plugin Root Fix & Release Docs

### PR DESCRIPTION
**Pull request by siegeon**
## Summary

- **PRISM Dashboard TUI** (`/prism-dashboard`) — live Textual terminal dashboard polling `prism-loop.local.md` with 8-step workflow table, agent roster, timing panel, gate alerts, and story info panel
- **CLI Snapshot mode** (`--snapshot`) — non-interactive ASCII dashboard output for embedding in Claude agent sessions
- **Plugin root resolution fix** — `_find_plugin_root()` sentinel-walk replaces hardcoded `parents[3]` depth, fixing breakage when run from the marketplace cache path
- **Pre-commit quality gate** — `.git/hooks/pre-commit` wired to `scripts/pre-commit`; fixes v2.4.0 gate never firing
- **validate-all path fix** — eliminates 234 false-positive documentation errors when run from repo root
- **Session-story-branch correlation** — state file tracks active git branch alongside session ID and story file
- **Activity hook** — `prism_activity_hook.py` records last-active timestamp for staleness detection
- **Release documentation** — v2.5.0 CHANGELOG entry, version bump, 22 AC-traced release doc tests

## Test plan

- [x] Full test suite: **117/117 passing** (`python -m pytest`)
- [x] Lint: **10.00/10** (`python -m pylint --recursive=y plugins/prism-devtools/tools/prism-cli/`)
- [x] Pre-commit gate passes (Phase 1 docs + Phase 2 portability)
- [x] TDD cycle complete — RED (20 failing) → GREEN (22/22 passing) for release docs story
- [ ] Manual: launch `/prism-dashboard` and verify TUI renders
- [ ] Manual: run `python prism-cli --snapshot` and verify ASCII output

## Stories on this branch

| Story | Status |
|-------|--------|
| PLAT-0000-prism-dashboard-tui | in_progress |
| PLAT-0000-precommit-gate-plugin-root-fix | done |
| PLAT-0000-session-story-branch-correlation | draft |
| PLAT-0000-prism-token-tracking | in_progress |
| PLAT-0000-release-documentation | done |

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: plugins/prism-devtools/tools/prism-cli/app.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        # Push to every widget on every tick — no reactive gating
+        try:
+            self.query_one(AgentRoster).update_state(self._state)
+            self.query_one(WorkflowTable).update_state(self._state)
+            self.query_one(StepDetail).update_state(self._state)
+            self.query_one(TimingPanel).update_state(self._state)
+            self.query_one(GatePanel).update_state(self._state)
+            self.query_one(StoryPanel).update_story(self._story)
+        except Exception:
+            pass  # Widgets not mounted yet during startup
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/tools/prism-cli/app.py:234]: > 🤖 **Claude** | code-reviewer, silent-failure-hunter

**[F1] Score: 92 — `except Exception: pass` swallows ALL widget errors on every poll tick**

This `except Exception: pass` block runs on every 2-second poll cycle for the entire application lifetime, not just during startup. If any widget's `update_state()` has a bug (`TypeError`, `AttributeError`, `KeyError`, etc.), the error is silently swallowed and the dashboard displays stale data indefinitely — with no log entry and no user-visible error.

Additionally, if `AgentRoster` succeeds but `StepDetail` throws, widgets 4-6 never get updated, creating an inconsistent UI state.

**Suggested fix:**
```python
from textual.css.query import NoMatches

try:
    self.query_one(AgentRoster).update_state(self._state)
    # ... other widgets ...
except NoMatches:
    pass  # Widgets not mounted yet during startup
except Exception:
    self.log.error(f"Widget update failed", exc_info=True)
```

File: plugins/prism-devtools/hooks/prism_activity_hook.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    content = _update_field(content, "last_activity", now)
+    if thought:
+        content = _update_field(content, "last_thought", thought)
+
+    try:
+        STATE_FILE.write_text(content, encoding="utf-8")
+    except (IOError, OSError):
+        pass
+
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/hooks/prism_activity_hook.py:124]: > 🤖 **Claude** | code-reviewer

**[F2] Score: 90 — Race condition: non-atomic read-modify-write on state file**

This hook fires on every `PostToolUse` event. When Claude makes parallel tool calls (which it frequently does), multiple activity hook instances run concurrently. Each reads the state file, modifies it in memory, and writes it back — a classic read-modify-write race. The second write overwrites the first's changes.

More critically, the stop hook (`prism_stop_hook.py`) also reads and writes the same state file. If an activity hook fires between the stop hook's read and write, one will clobber the other, potentially losing step advancement state.

**Suggested fix:** Either:
1. Use a file lock (`filelock` package or platform-specific locking) around the read-modify-write cycle
2. Write `last_activity` and `last_thought` to a separate file (e.g., `.claude/prism-activity.json`) to avoid contention with the state file

File: plugins/prism-devtools/skills/jira/scripts/jira_fetch.py [hunk 3]
```diff
@@ -27,7 +27,17 @@
...
-ENV_FILE_PATH = Path("C:/Dev/.prism/plugins/prism-devtools/.env")
+
+def _find_plugin_root() -> Path:
+    """Walk up from __file__ to find the plugin root (contains core-config.yaml)."""
+    current = Path(__file__).resolve().parent
+    while current != current.parent:
+        if (current / "core-config.yaml").exists():
+            return current
+        current = current.parent
+    raise FileNotFoundError("Could not find plugin root (no core-config.yaml in any ancestor)")
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/skills/jira/scripts/jira_fetch.py:38]: > 🤖 **Claude** | code-reviewer

**[F3] Score: 88 — `_find_plugin_root()` crashes at module import if `core-config.yaml` absent**

`PLUGIN_ROOT = _find_plugin_root()` executes at module scope (import time). If `core-config.yaml` doesn't exist in any ancestor directory (e.g., running from a test harness, CI, or any context outside the plugin tree), the function raises `FileNotFoundError` and the entire module fails to import.

This same pattern exists in `jira_search.py`, `prism_approve.py`, `prism_reject.py`, and `setup_prism_loop.py`.

**Suggested fix:** Make the call lazy — execute inside `main()` or at point of use rather than at module scope. For test contexts, fall back to `Path(__file__).resolve().parents[3]`.

File: plugins/prism-devtools/hooks/prism_activity_hook.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    def _update_field(text: str, key: str, value: str) -> str:
+        pattern = rf"^{key}:\s*.*$"
+        replacement = f'{key}: "{value}"'
+        if re.search(pattern, text, re.MULTILINE):
+            return re.sub(pattern, lambda m: replacement, text, flags=re.MULTILINE)
+        # Add before closing ---
+        return re.sub(r"^(---)$", lambda m: f"{replacement}\n---", text, count=1, flags=re.MULTILINE)
+
+    content = _update_field(content, "last_activity", now)
+    if thought:
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/hooks/prism_activity_hook.py:109]: > 🤖 **Claude** | code-reviewer

**[F4] Score: 85 — `_update_field` regex inserts before opening `---` instead of closing**

When a field doesn't exist in the frontmatter, `re.sub(r"^(---)$", ..., count=1)` matches the **opening** `---` delimiter (line 1 of the file), not the closing one. This inserts the new field *above* the frontmatter block rather than inside it.

YAML frontmatter is delimited by `---` on both the first and last lines. With `count=1`, the regex matches the first occurrence.

**Suggested fix:**
```python
# Insert before the closing --- by matching the last occurrence
return re.sub(r"\n(---)$", lambda m: f"\n{replacement}\n---", text, count=1, flags=re.MULTILINE)
```
Or use a different approach like splitting on frontmatter boundaries.

File: plugins/prism-devtools/tools/prism-cli/validate.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        result.check(
+            row_count == 8,
+            f"WorkflowTable has {row_count} rows (expected 8)",
+        )
+
+        # Check the current step's status cell contains the right marker
+        if row_count == 8:
+            wf_rows = wf_data_table.ordered_rows
+            wf_cols = wf_data_table.ordered_columns
+            status_col_key = wf_cols[9].key  # "Status" column (index 9 after Skills added)
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/tools/prism-cli/validate.py]: > 🤖 **Claude** | code-reviewer, comment-analyzer

**[F5] Score: 85 — Hardcoded magic column indices `[9]` and `[4]` for column lookup**

`wf_cols[9]` and `agent_cols[4]` are brittle magic numbers. The comment even acknowledges this: `"index 9 after Skills added"` — meaning this was already updated once when a column was added. If anyone adds or reorders a column in `WorkflowTable` or `AgentRoster`, the validation silently checks the wrong column, producing false PASSes or FAILs in the acceptance gate.

**Suggested fix:**
```python
status_col_key = next(
    col.key for col in wf_cols if "Status" in str(col.label)
)
agent_state_col = next(
    col.key for col in agent_cols if "State" in str(col.label)
)
```

File: plugins/prism-devtools/hooks/prism_activity_hook.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        content = _update_field(content, "last_thought", thought)
+
+    try:
+        STATE_FILE.write_text(content, encoding="utf-8")
+    except (IOError, OSError):
+        pass
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/hooks/prism_activity_hook.py]: > 🤖 **Claude** | silent-failure-hunter

**[F6] Score: 82 — State file write `IOError` silently swallowed**

The dashboard's staleness detector relies on `last_activity` being updated on every tool call. If writes fail silently here (disk full, file locked by concurrent process), the dashboard will show "STALE" even though the agent is actively working. The user could mistakenly cancel a healthy session.

**Suggested fix:**
```python
except (IOError, OSError) as e:
    print(f"Warning: Could not update PRISM state file: {e}", file=sys.stderr)
```

File: plugins/prism-devtools/tools/prism-cli/models.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        return max(0, self.total_tokens - self.step_tokens_start)
+
+    @property
+    def step_history_parsed(self) -> list[dict]:
+        """List of {i, d, t} for each completed step."""
+        try:
+            import json
+            return json.loads(self.step_history)
+        except Exception:
+            return []
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/tools/prism-cli/models.py:57]: > 🤖 **Claude** | silent-failure-hunter

**[F7] Score: 82 — `except Exception: return []` too broad in `step_history_parsed`**

If `self.step_history` contains corrupt JSON, this returns `[]` silently. The timing/token calculations then evaluate to 0. The user sees "0 tokens" for all completed steps with no indication the data is corrupt.

A `TypeError` or `AttributeError` from an unexpected code path would also be swallowed here.

**Suggested fix:**
```python
except json.JSONDecodeError:
    return []
```
Only catch the expected JSON parsing error.

File: plugins/prism-devtools/hooks/prism_stop_hook.py [hunk 8]
```diff
@@ -764,11 +909,40 @@ def main():
...
     # Update state to next step
     updates = {
         "current_step": next_step_id,
         "current_step_index": next_index,
-        "last_activity": datetime.now().isoformat(),
+        "last_activity": now_ts.isoformat(),
+        "step_started_at": now_ts.isoformat(),
+        "step_tokens_start": str(usage["total_tokens"]),
+        "step_transcript_line": str(usage["total_lines"]),
+        "step_history": json.dumps(history),
```
Comment (danpuzon-resolve) [inline plugins/prism-devtools/hooks/prism_stop_hook.py:947]: > 🤖 **Claude** | code-reviewer

**[F8] Score: 80 — `step_history` JSON written unquoted to frontmatter**

`update_state_file` writes `step_history: [{"i": 0, "d": 45, "t": 1000}]` without surrounding quotes. This works today because `split(":", 1)` extracts the key correctly and `json.loads()` handles the raw array string.

However, `parsing.py`'s `update_state_field` (line 229) does quote non-numeric values — creating an inconsistency between the two update paths. There's no round-trip test verifying that `step_history` survives write-then-read through both paths.

**Suggested fix:** Either quote JSON values in `update_state_file`:
```python
value_str = f'"{json.dumps(value)}"' if isinstance(value, (list, dict)) else str(value)
```
Or add a round-trip test covering `step_history` through both the stop hook and parsing module paths.

**Review by danpuzon-resolve — CHANGES_REQUESTED** [review]
> 🤖 **Claude Code Review**

## Summary

Reviewed 51 changed files (+5,416/-36) across the PRISM Dashboard TUI, CLI snapshot mode, plugin root fix, and release docs. The TUI is well-structured with good test coverage for the parsing layer. Two parallel review pipelines (4 agents) identified **8 actionable issues** and **7 summary-level observations**.

### Issues by Severity

| # | Score | Severity | File | Issue |
|---|-------|----------|------|-------|
| F1 | 92 | High | `app.py:222-233` | `except Exception: pass` in `_poll_state` swallows all widget errors |
| F2 | 90 | High | `prism_activity_hook.py:65-110` | Race condition: non-atomic read-modify-write on state file |
| F3 | 88 | High | 5 files | `_find_plugin_root()` crashes at import time if sentinel missing |
| F4 | 85 | Medium | `prism_activity_hook.py:100-103` | `_update_field` regex inserts before opening `---` |
| F5 | 85 | Medium | `validate.py:179,218` | Hardcoded magic column indices for column lookup |
| F6 | 82 | Medium | `prism_activity_hook.py:106-109` | State file write IOError silently swallowed |
| F7 | 82 | Medium | `models.py:53-57` | `except Exception` too broad in `step_history_parsed` |
| F8 | 80 | Medium | `prism_stop_hook.py:708-716` | `step_history` JSON written unquoted to frontmatter |

### Summary-Only Observations (score < 80)

- **F9 (75):** `_find_plugin_root()` is duplicated across 5 scripts — extract to shared utility
- **F10 (72):** Misleading comment in `prism_stop_hook.py:32` describes validation as `"red"/"green"/None` but actual values are `story_complete`, `plan_coverage`, `red_with_trace`, `green`, `green_full`
- **F11 (72):** `models.py:101` comment says "mirrors prism_stop_hook.py WORKFLOW_STEPS" but schema is completely different (dataclass vs 6-tuple, different fields)
- **F12 (70):** `_fmt_duration` and `_fmt_tokens` duplicated across `agent_roster.py`, `workflow_table.py`, `snapshot.py`, and `app.py`
- **F13 (70):** `AGENTS` list duplicated between `snapshot.py:77` and `agent_roster.py` — import or extract
- **F14 (65):** `isdigit()` quoting guard in `parsing.py:227` doesn't handle negative integers
- **F15 (60):** Reference to "POSIX-path bug" in `app.py:188` with no context or issue link

### What Looks Good

- `_find_plugin_root()` sentinel-walk correctly replaces fragile `parents[3]` depth hack
- `snapshot.py` incremental transcript reading with `last_good` offset handles partial writes well
- Test suite has thorough AC traceability headers with good edge case coverage
- `validate-all.py` fix for `--root .` path resolution bug is well-documented
- Windows junction detection via `resolve()` comparison is a solid cross-platform approach

### Scoring Transparency

- **4 agents** ran in parallel across 2 pipelines
- **35 raw findings** across all agents → **23 unique after dedup** → **8 posted** (score ≥ 80)
- **16 pre-existing findings** filtered (not in this PR's diff)

File: docs/stories/PLAT-0000-precommit-gate-plugin-root-fix.story.md
Comment (danpuzon-resolve) [inline docs/stories/PLAT-0000-precommit-gate-plugin-root-fix.story.md:1]: Wonder if we should .gitignore these docs/stories files? They are actually kinda cool to see but may get out of hand over time as we iterate

**Review by danpuzon-resolve — APPROVED** [review]
Agentic review left some comments that sound legit, consider addressing or we can iterate and fix later (or ignore if they aren't relevant)
